compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -534,23 +534,45 @@ private BType getFiniteTypeMatchWithIntLiteral(BLangLiteral literalExpr, BFinite
         if (intLiteralType != symTable.noType) {
             return intLiteralType;
         }
+        
         int typeTag = getPreferredMemberTypeTag(finiteType);
         if (typeTag == TypeTags.NONE) {
             return symTable.intType;
         }
+        
         if (literalAssignableToFiniteType(literalExpr, finiteType, typeTag)) {
             BType type = symTable.getTypeFromTag(typeTag);
             setLiteralValueForFiniteType(literalExpr, type, data);
             literalExpr.value = String.valueOf(literalValue);
             return type;
         }
+
         // Handle out of range ints
         if (literalValue instanceof Double) {
             return symTable.floatType;
         }
         if (literalValue instanceof String) {
             return symTable.decimalType;
         }
+        
+        BType matchingType = null;
+        for (int tag = TypeTags.FLOAT; tag <= TypeTags.DECIMAL; tag++) {
+            BType type = symTable.getTypeFromTag(tag);
+            for (BLangExpression memType : finiteType.getValueSpace()) {
+                if (memType.getBType().tag == tag) {
+                    matchingType = type;
+                    break;
+                }
+            }
+
+            if (matchingType != null) {
+                break;
```

Comment:
We can move the loops to a helper function and return instead of breaking to avoid this check.